projects/eslint-plugin/rules/no-tui-assert-without-ng-dev-mode.js

```diff
@@ -0,0 +1,152 @@
+    meta: {
+        type: 'problem',
+        docs: {description: ERROR_MESSAGE},
+        messages: {
+            [MESSAGE_ID]: ERROR_MESSAGE,
+        },
+        fixable: 'code',
+    },
+    create(context) {
+        return {
+            /**
+             * @param node {import('@types/estree').ExpressionStatement}
+             */
+            'CallExpression[callee.object.name="tuiAssert"][callee.property.name="assert"]'(
+                node,
+            ) {
+                const ancestors = context.getAncestors();
+                const logicalExpressions = getLogicalExpressions(ancestors);
+
+                /**
+                 * There are 3 cases, the goal is to find a required identifier in any of them.
+                 * 1. ngDevMode && tuiAssert.assert()
+                 * 2. if (ngDevMode && isUser) { tuiAssert.assert() }
+                 * 3. ngDevMode ? tuiAssert.assert() : null
+                 */
+                if (
+                    validateLogicalExpressions(logicalExpressions) ||
+                    validateIfExpressions(ancestors) ||
+                    validateConditionalExpressions(ancestors)
+                ) {
+                    return;
+                }
+
+                context.report({
+                    node,
+                    messageId: MESSAGE_ID,
+                    /**
+                     * This fixer is pretty straightforward.
+                     * The best solution requires complicated logic,
+                     * handling different types of expressions and combinations of them.
+                     */
+                    fix: fixer => [
+                        fixer.insertTextBefore(node, `(${requiredIdentifierName} && `),
+                        fixer.insertTextAfter(node, `)`),
+                    ],
+                });
+            },
+        };
+    },
+};
+
+/**
+ * Filters the provided ancestors by a LogicalExpression type.
+ * @param ancestors {import('@types/estree').Node[]}
+ * @returns {import('@types/estree').LogicalExpression[]}
+ */
+const getLogicalExpressions = ancestors => {
+    return ancestors.filter(({type}) => type === 'LogicalExpression');
+};
+
+/**
+ * Tries to find a required identifier in the provided logical expression.
+ * @param expression {import('@types/estree').LogicalExpression}
+ * @returns {import('@types/estree').Identifier | null}
+ */
+const getRequiredIdentifierFromLogicalExpression = expression => {
+    let requiredIdentifier = null;
+
+    traverseLogicalExpression(expression, leaf => {
+        if (leaf.name === requiredIdentifierName) {
+            requiredIdentifier = leaf;
+        }
+    });
+
+    return requiredIdentifier;
+};
+
+/**
+ * Validates that a required identifier is presented in the provided expressions.
+ * Handles `ngDevMode && tuiAssert.assert()` expression.
+ * @param expressions {import('@types/estree').LogicalExpression[]}
+ * @returns {boolean}
+ */
+const validateLogicalExpressions = expressions =>
```

Comment:
can we separate to another files as utils?

Reply:
sure, as you wish

Reply:
can you please explain what you want to move exactly? the `validateLogicalExpressions` function is very short and pretty specific

Reply:
`can be moved` sure

<img width="659" alt="image" src="https://user-images.githubusercontent.com/12021443/230793045-2ea9ae19-4dc9-4c31-84fe-131fc5fe1b0c.png">